### ISSUE TITLE
error: invalid use of incomplete type ‘class std::__make_unsigned_selector<__int128 unsigned, false, false>’

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#1902]
From a [PPA build of the development branch in Ubuntu 18.04 on amd64](https://launchpadlibrarian.net/520401111/buildlog_ubuntu-bionic-amd64.macaulay2_1.17.2.1+git202101300204-0ppa202101300214~ubuntu18.04.1_BUILDING.txt.gz):

```shell
g++ -pipe -Wall -Wno-shadow -Wcast-qual -Wno-sign-conversion -Wno-sign-compare -Wno-parentheses -Wno-sign-compare  -fno-stack-protector -Wuninitialized -I/usr/include  -Wno-error=pragmas -std=gnu++14 -g -fdebug-prefix-map=/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204=. -fstack-protector-strong -Wformat -Werror=format-security -g3 -O2  -Wno-unused-local-typedefs -Wno-unknown-pragmas -std=c++17 -Wno-mismatched-tags -I. -DNDEBUG -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/usr-host/include  -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/usr-host/include  -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202101300204/M2/usr-host/include  -Wdate-time -D_FORTIFY_SOURCE=2 -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -I../d -I./../c -Wno-fatal-errors  -c -Wno-cast-qual VectorArithmetic.cpp -o VectorArithmetic.o
In file included from MemoryBlock.hpp:8:0,
                 from VectorArithmetic.hpp:6,
                 from VectorArithmetic.cpp:1:
/usr/include/tbb/tbb.h:21:154: note: #pragma message: TBB Warning: tbb.h contains deprecated functionality. For details, please see Deprecated Features appendix in the TBB reference manual.
 #pragma message("TBB Warning: tbb.h contains deprecated functionality. For details, please see Deprecated Features appendix in the TBB reference manual.")
                                                                                                                                                          ^
In file included from /usr/include/c++/7/bits/move.h:54:0,
                 from /usr/include/c++/7/bits/nested_exception.h:40,
                 from /usr/include/c++/7/exception:143,
                 from /usr/include/c++/7/new:40,
                 from /usr/include/gc/gc_allocator.h:45,
                 from ./newdelete.hpp:10,
                 from NCAlgebras/Range.hpp:4,
                 from VectorArithmetic.hpp:4,
                 from VectorArithmetic.cpp:1:
/usr/include/c++/7/type_traits: In instantiation of ‘struct std::make_unsigned<__int128 unsigned>’:
/usr/include/givaro/modular-int64.h:46:60:   required from ‘class Givaro::Modular<long int, __int128 unsigned>’
/usr/include/givaro/modular-int64.inl:32:36:   required from here
/usr/include/c++/7/type_traits:1817:62: error: invalid use of incomplete type ‘class std::__make_unsigned_selector<__int128 unsigned, false, false>’
     { typedef typename __make_unsigned_selector<_Tp>::__type type; };
                                                              ^~~~
/usr/include/c++/7/type_traits:1774:11: note: declaration of ‘class std::__make_unsigned_selector<__int128 unsigned, false, false>’
     class __make_unsigned_selector;
           ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/7/type_traits: In instantiation of ‘struct std::make_unsigned<__int128>’:
/usr/include/givaro/modular-int64.h:46:60:   required from ‘class Givaro::Modular<long int, __int128>’
/usr/include/givaro/modular-int64.inl:36:35:   required from here
/usr/include/c++/7/type_traits:1817:62: error: invalid use of incomplete type ‘class std::__make_unsigned_selector<__int128, false, false>’
     { typedef typename __make_unsigned_selector<_Tp>::__type type; };
                                                              ^~~~
/usr/include/c++/7/type_traits:1774:11: note: declaration of ‘class std::__make_unsigned_selector<__int128, false, false>’
     class __make_unsigned_selector;
           ^~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: warning: unrecognized command line option ‘-Wno-mismatched-tags’
Makefile.common:32: recipe for target 'VectorArithmetic.o' failed
make[4]: *** [VectorArithmetic.o] Error 1
```

**Comment by d-torrance** [comment]
This also happened on [arm64](https://launchpadlibrarian.net/520402112/buildlog_ubuntu-bionic-arm64.macaulay2_1.17.2.1+git202101300204-0ppa202101300214~ubuntu18.04.1_BUILDING.txt.gz) and [s390x](https://launchpadlibrarian.net/520401065/buildlog_ubuntu-bionic-s390x.macaulay2_1.17.2.1+git202101300204-0ppa202101300214~ubuntu18.04.1_BUILDING.txt.gz), also in Ubuntu 18.04.

**Comment by d-torrance** [comment]
The 32-bit architectures (i386 and armhf) compiled just fine.  Both are happily building documentation right now.

**Comment by d-torrance** [comment]
As with #1901, it looks like compiling with GNU extensions (`-std=gnu++17` instead of `-std=c++17`) may fix this.  See for example https://stackoverflow.com/questions/38806435/build-proxygen-under-ubuntu-16-04.  I'll report back after the next PPA build.

**Comment by d-torrance** [comment]
I confirmed in an Ubuntu 18.04 chroot that compiling with `-std-gnu++17` fixes this.

**Comment by DanGrayson** [comment]
> I confirmed in an Ubuntu 18.04 chroot that compiling with `-std-gnu++17` fixes this.

@mikestillman  -- this may interest you.

**Comment by mikestillman** [comment]
The main reason I had suggested/switched to -std=std++17 is so that we would not be dependent on gnu extensions (e.g. for intel c++ compiler).  But it seems like the packages we are using require these gnu extensions, even for our own compilation.  I don't mind using gnu++17, as I'm always compiling on gcc or clang, but I wonder if this means we can't compile on intel c++? (It would be nice to not rule that out, although I currently don't have access to this compiler).

**Comment by d-torrance** [comment]
Supposedly the Intel compiler supports (many of) the GNU extensions: https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compatibility-and-portability/gcc-compatibility-and-interoperability.html

**Comment by d-torrance** [comment]
This was closed by #1923.
